### ISSUE TITLE
Minio in gateway mode with caching fails to retrieve file

### ISSUE THREAD

**Issue by aronneagu** [minio/minio#13292]
When using minio in gateway mode with cache enabled, if there is file in the minio cache, and that file is being overwritten in the backend (Azure storage) and retrieved via minio, the first attempt at getting the new version will fail (and remove the file from cache). Second attempt to fetch the file will succeed and populate the cache accordingly

## Expected Behavior
When minio is retrieving the new version of the file, it should not error

## Possible Solution
One solution was to add `--disable-multipart` to the command when uploading the file

## Steps to Reproduce (for bugs)
```
#!/bin/sh
mc alias set minio-caching http://host.mycompany.com:9091 bcminiotest "secret_key"
mc mb minio-caching bcminiotest
rm testfile*
dd if=/dev/urandom of=testfile.input.1 bs=1k count=20000
dd if=/dev/urandom of=testfile.input.2 bs=1k count=20000
md5sum testfile.input.*

mc rm minio-caching/bcminiotest/testfile
# Send file to Minio
mc cp testfile.input.1 minio-caching/bcminiotest/testfile
mc cp minio-caching/bcminiotest/testfile testfile.output.1

# Send file 2 to Minio
mc cp testfile.input.2 minio-caching/bcminiotest/testfile
mc cp minio-caching/bcminiotest/testfile testfile.output.2-1
mc cp minio-caching/bcminiotest/testfile testfile.output.2-2

md5sum testfile*
```
Output
```
mc: <ERROR> Failed to copy `http://host.mycompany.com:9091/bcminiotest/testfile`. At least one of the pre-conditions you specified did not hold
178e1f9a59b0b7d197aa509243f027a2  testfile.input.1
8f7d21b8c6fcfe41afc5308ca2160535  testfile.input.2
178e1f9a59b0b7d197aa509243f027a2  testfile.output.1
8f7d21b8c6fcfe41afc5308ca2160535  testfile.output.2-2
```
Would have expected to see testfile.output.2-1, but instead got the error above

## Context
We have multiple teams accessing data in Azure Storage. This increases the bandwidth usage, latency and cost associated with fetching data from Azure. To reduce all that, we want to run a local minio instance (in the office) that acts as a cache layer between the office and Azure.

## Your Environment
* Server setup and configuration:
```
> minio --version
minio version RELEASE.2020-10-18T21-54-12Z
> cat /etc/systemd/system/minio-caching
...
EnvironmentFile=/etc/default/minio-caching
ExecStart=/usr/local/bin/minio gateway azure --address :9091
...
> cat /etc/default/minio-caching
MINIO_ACCESS_KEY="bcminiotest"
MINIO_SECRET_KEY=
MINIO_CACHE=on
MINIO_CACHE_DRIVES=/mnt/bcminiocaching
MINIO_CACHE_QUOTA=90
MINIO_CACHE_AFTER=0
MINIO_CACHE_WATERMARK_LOW=70
MINIO_CACHE_WATERMARK_HIGH=90
MINIO_CACHE_RANGE=off
MINIO_PROMETHEUS_AUTH_TYPE="public"
> uname -a 
Linux BC-PC34 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by poornas** [comment]
While this looks unexpected, the cache is working as designed.  `mc cp` does more under the hood than a regular `GET`, it first performs a `HEAD` and then `GET` with an If-ETag-Match option using metadata retrieved in the previous `HEAD` call. 

The first HEAD returned ETag of the stale entry in the cache and was used to validate with Azure backend during the `GET`. Here it is the Azure backend that returns the ` At least one of the pre-conditions you specified did not hold` error. If you were to use a different client, you would not see this issue

**Comment by aronneagu** [comment]
Hi Poornas,

If I run the above script and add `--multipart-disable` to `mc cp <local> <minio-caching>`, then retrieving the files does not error.
The output of that would be
```
7b49c2a91ef8089dc46bf8f84b96c9d0  testfile.input.1
181b6ab511a1b0ffe0bec10354227766  testfile.input.2
7b49c2a91ef8089dc46bf8f84b96c9d0  testfile.output.1
181b6ab511a1b0ffe0bec10354227766  testfile.output.2-1
181b6ab511a1b0ffe0bec10354227766  testfile.output.2-2
```
As you can see, the first retrieval of the new version of the file (testfile.output.2-1) works correctly, and even though in the cache it contains the file with hash 7b49c2a...., minio reaches out to Azure, notices the file in the backend (Azure storage) has changed, it fetches the new version (hash 181b6a) and serves that to the requestor. This is what I would expect to be the correct behavior

My dilemma is why this is failing if the file is uploaded in parts?

**Comment by poornas** [comment]
objects uploaded via multipart are cached on GET, not on upload. Because of this, the first `GET` caches it, however the overwrite did not invalidate the cache entry because it is a multipart object. 

Note that when you specify --disable-multipart, `mc cp` would upload the file as a single part, and the cache saves it locally on upload - thus allowing overwrites to replace the cache entry.

**Comment by harshavardhana** [comment]
> objects uploaded via multipart are cached on GET, not on upload. Because of this, the first `GET` caches it, however the overwrite did not invalidate the cache entry because it is a multipart object.

@poornas should we not just overwrite anyways if the upload is a single part?

**Comment by aronneagu** [comment]
How come multipart uploads are not cached on upload? What is preventing that?

**Comment by harshavardhana** [comment]
> How come multipart uploads are not cached on upload? What is preventing that?

It is a design decision, there are lots of complexities involved.

**Comment by aronneagu** [comment]
We are using the python api sdk to upload files to a minio instance witch caching enabled. 
As the multipart upload does not update the cache, we are left to use single part upload.
However from the documentation (https://docs.min.io/docs/python-client-api-reference.html#put_object) the max file size is 5Gb for single part upload, while our files are up to 10Gb in size.

Is there a workaround for this?

**Comment by klauspost** [comment]
@aronneagu We will fix the missing cache invalidation on multipart uploads when we get to it.

**Comment by poornas** [comment]
> @poornas should we not just overwrite anyways if the upload is a single part?

That already happens for single part - the issue here is cache and backend have different copies of the object due to object being uploaded as multipart(which is cached on GET). 

His order of operations is multipart PUT, GET, multipart overwrite, GET - this fails on the last `mc cp gw/bucket/obj`  because mc performed a `HEAD` which would have returned `ETag` of the cached entry,  then mc sends a `GET` with If-Match using the ETag returned by cache.

**Comment by harshavardhana** [comment]
> > @poornas should we not just overwrite anyways if the upload is a single part?
> 
> That already happens for single part - the issue here is cache and backend have different copies of the object due to object being uploaded as multipart(which is cached on GET).
> 
> His order of operations is multipart PUT, GET, multipart overwrite, GET - this fails on the last `mc cp gw/bucket/obj` because mc performed a `HEAD` which would have returned `ETag` of the cached entry, then mc sends a `GET` with If-Match using the ETag returned by cache.

Ah i see there is no fix for this AFAICS then

**Comment by harshavardhana** [comment]
This is fixed by #13507 @poornas right?

**Comment by poornas** [comment]
> This is fixed by #13507 @poornas right?

yes
